pkg/kubelet/server.go

```diff
@@ -587,80 +595,160 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+		// at supporting old clients that don't generate connection IDs.
+		// If there are concurrent new connections, it's possible that 1
+		// connection gets streams whose IDs are not consecutive (e.g. 5
+		// and 9 instead of 5 and 7). In this case, it's likely that
+		// the client will be unable to forward any more ports until
+		// the `kubectl port-forward` command is terminated and restarted.
+		connectionID = strconv.Itoa(int(streamID) / 4)
+		glog.V(5).Infof("Automatically assigning connection ID %d from stream ID %d", connectionID, streamID)
+	}
+	return connectionID
+}
+
+// ServePortForward handles a port forwarding request.
+func ServePortForward(w http.ResponseWriter, req *http.Request, portForwarder PortForwarder, podName string, uid types.UID, idleTimeout time.Duration) {
+	streamChan := make(chan httpstream.Stream, 1)
+
+	glog.V(5).Infof("Upgrading port forward response")
+	upgrader := spdy.NewResponseUpgrader()
+	conn := upgrader.UpgradeResponse(w, req, portForwardStreamReceived(streamChan))
 	if conn == nil {
 		return
 	}
 	defer conn.Close()
-	conn.SetIdleTimeout(s.host.StreamingConnectionIdleTimeout())
 
-	var dataStreamLock sync.Mutex
-	dataStreamChans := make(map[string]chan httpstream.Stream)
+	glog.V(5).Infof("Setting port forwarding streaming connection idle timeout to %v", idleTimeout)
+	conn.SetIdleTimeout(idleTimeout)
 
+	var streamLock sync.Mutex
+	streamChans := make(map[string]chan httpstream.Stream)
+
+	glog.V(5).Infof("Waiting for port forward streams")
 Loop:
 	for {
 		select {
 		case <-conn.CloseChan():
+			glog.V(5).Infof("Upgraded connection closed")
 			break Loop
 		case stream := <-streamChan:
-			streamType := stream.Headers().Get(api.StreamType)
-			port := stream.Headers().Get(api.PortHeader)
-			dataStreamLock.Lock()
-			switch streamType {
-			case "error":
-				ch := make(chan httpstream.Stream)
-				dataStreamChans[port] = ch
-				go waitForPortForwardDataStreamAndRun(kubecontainer.GetPodFullName(pod), uid, stream, ch, s.host)
-			case "data":
-				ch, ok := dataStreamChans[port]
-				if ok {
-					ch <- stream
-					delete(dataStreamChans, port)
-				} else {
-					glog.Errorf("Unable to locate data stream channel for port %s", port)
-				}
-			default:
-				glog.Errorf("streamType header must be 'error' or 'data', got: '%s'", streamType)
-				stream.Reset()
+			connectionID := getPortForwardStreamConnectionID(stream)
+			glog.V(5).Infof("(%s) Received new stream", connectionID)
+
+			streamLock.Lock()
+
+			ch, ok := streamChans[connectionID]
+			if !ok {
+				glog.V(5).Infof("(%s) connection ID not in streamChans", connectionID)
+
+				glog.V(5).Infof("(%s) creating stream channel", connectionID)
+				ch = make(chan httpstream.Stream)
+				streamChans[connectionID] = ch
+
+				glog.V(5).Infof("(%s) starting handlePortForwardStreams", connectionID)
+				go handlePortForwardStreams(podName, uid, ch, portForwarder, connectionID, streamChans, streamLock)
```

Comment:
I'd prefer the timeout and cleanup handling be done in this function, rather than passing the map, lock and connectionID into `handlePortForwardStreams`... trying to think of cleanest way to do that